crates/bevy_transform/src/systems.rs

```diff
@@ -1,75 +1,84 @@
+    parent_query: Query<&Parent>,
     children_query: Query<(&Children, Changed<Children>), (With<Parent>, With<GlobalTransform>)>,
 ) {
-    for (children, transform, transform_changed, mut global_transform, entity) in
-        root_query.iter_mut()
-    {
-        let mut changed = transform_changed;
-        if transform_changed {
-            *global_transform = GlobalTransform::from(*transform);
-        }
+    root_query.par_for_each_mut(
+        1,
+        |(entity, children, transform, changed, mut global_transform)| {
+            if changed {
+                *global_transform = GlobalTransform::from(*transform);
+            }
 
-        if let Some((children, changed_children)) = children {
-            // If our `Children` has changed, we need to recalculate everything below us
-            changed |= changed_children;
             for child in children.iter() {
                 let _ = propagate_recursive(
                     &global_transform,
-                    &mut transform_query,
+                    &transform_query,
+                    &parent_query,
                     &children_query,
-                    *child,
                     entity,
+                    *child,
                     changed,
                 );
             }
-        }
-    }
+        },
+    );
 }
 
 fn propagate_recursive(
     parent: &GlobalTransform,
-    transform_query: &mut Query<(
-        &Transform,
-        Changed<Transform>,
-        &mut GlobalTransform,
-        &Parent,
-    )>,
+    transform_query: &Query<(&Transform, Changed<Transform>, &mut GlobalTransform), With<Parent>>,
+    parent_query: &Query<&Parent>,
     children_query: &Query<(&Children, Changed<Children>), (With<Parent>, With<GlobalTransform>)>,
-    entity: Entity,
     expected_parent: Entity,
+    entity: Entity,
     mut changed: bool,
     // We use a result here to use the `?` operator. Ideally we'd use a try block instead
 ) -> Result<(), ()> {
-    let global_matrix = {
-        let (transform, transform_changed, mut global_transform, child_parent) =
-            transform_query.get_mut(entity).map_err(drop)?;
-        // Note that for parallelising, this check cannot occur here, since there is an `&mut GlobalTransform` (in global_transform)
+    if let Ok(actual_parent) = parent_query.get(entity) {
         assert_eq!(
-            child_parent.0, expected_parent,
+            actual_parent.0, expected_parent,
             "Malformed hierarchy. This probably means that your hierarchy has been improperly maintained, or contains a cycle"
         );
+    } else {
+        panic!("Propagated child for {:?} has no Parent component!", entity);
+    }
+
+    // SAFE: With the check that each child has one and only one parent, each child must be globally unique within the
+    // hierarchy. Because of this, it is impossible for this query to have aliased mutable access to the same GlobalTransform.
```

Comment:
This proof could use some expansion imo. E.g.

The above check confirms that we are in the only path to reach this entity, since any other path would have found that the parent did not match as expected and panicked.
Therefore, this is the only time in this system that the `GlobalTransform` component of this entity is accessed, so this does not alias.

I don't think this is complete, but I think the current proof is too hand-wavy.

Reply:
I expanded this with diagrams for demonstration. Is this better?